src/docs/testing/best-practices.md

```diff
@@ -0,0 +1,174 @@
+`saveLayer()` behind the scenes, which can be an expensive operation.
+
+{{site.alert.note}}
+**Why is savelayer expensive?**<br>
+Calling `saveLayer()` allocates an offscreen buffer. Drawing content
+into the offscreen buffer may trigger render target switches that
+are particularly slow in older GPUs.
+{{site.alert.end}}
+
+Some general rules when applying specific effects:
+
+* Use the [**Opacity**]({{site.api}}/flutter/widgets/Opacity-class.html)
+  widget only when necessary, but if you must use it, add a
+  RepaintBoundary right below Opacity in the widget tree to prevent
+  unnecessary calls to `saveLayer()`.  See [Transparent
+  image]({{site.api}}/flutter/widgets/Opacity-class.html#transparent-image)
+  in the Opacity API page for an example of applying opacity directly
+  to an image, which is faster than using the Opacity widget.
+* **Clipping** doesn’t call `saveLayer()` (unless explicitly requested with
+  `Clip.antiAliasWithSaveLayer()`) so these operations aren’t as expensive
+  as Opacity, but clipping is still costly, so use with caution. By default,
+  clipping is disabled (`Clip.none`), so you must explicitly enable it when
+  needed. 
+
+Other widgets that may trigger `saveLayer()` and are potentially costly:
+
+* [ShaderMask]({{site.api}}/flutter/widgets/ShaderMask-class.html)
+* [ColorFilter]({{site.api}}/flutter/dart-ui/ColorFilter-class.html)
+* [Chip]({{site.api}}/flutter/material/Chip-class.html)&mdash;may
+  cause call to `saveLayer()` if `disabledColorAlpha != 0xff`
+* [Text]({{site.api}}/flutter/widgets/Text-class.html)&mdash;may
+  cause call to `saveLayer()` if there's an `overflowShader`
+
+Ways to avoid calls to `saveLayer()`:
+
+* To implement fading in an image, consider using the FadeInImage widget,
+  which applies a gradual opacity using the GPU’s fragment shader.
+  For more information, see
+  [Opacity]({{site.api}}/flutter/widgets/Opacity-class.html#transparent-image).
+  ([PENDING] not released yet)
+* To create a rectangle with rounded corners, instead of applying a
+  clipping rectangle, consider using the `borderRadius` property offered
+  by many of the widget classes.
+
+### Render grids and lists lazily
+
+Use the lazy methods, with callbacks, when building large grids or lists.
+That way only the visible portion of the screen is built at startup time.
+
+Also see:
+
+* [Working with long lists](/cookbook/lists/long-lists) in the
+  [Cookbook](/cookbook)
+* [Creating a ListView that loads one page at a
+  time](https://medium.com/saugo360/flutter-creating-a-listview-that-loads-one-page-at-a-time-c5c91b6fabd3)
+  by AbdulRahman AlHamali
+* [Listview.builder]({{site.api}}/flutter/widgets/ListView/ListView.builder.html) API
+
+###  Build and display frames in 16ms
+
+A frame should be built and rendered in 16ms (1/60th of a second) _or less_
+for a 60Hz display.  Note that means built in 8ms or less,
+and rendered in 8ms or less, for a total of 16ms or less.
```

Comment:
Since there are two separate threads for building and rendering, we actually can have 16ms for building and 16ms for rendering. @chinmaygarde : please correct me if I'm wrong.

Reply:
From discussion in the original doc for this article, this would introduce a frame of latency.

Reply:
Ah, yes, it will introduce latency. I thought it was only about janky frames. We probably should explicitly clarify that: if latency is a concern, go with total 16ms limit; if the only concern is the missing frames (or fps/jankyness/smoothness), then 16ms for each of the build and render stage is Ok.